packages/sdk/src/types/runtime/behaviors/action.ts

```diff
@@ -24,13 +29,25 @@ export class DiscreteAction {
 	private handlers: ActionHandlers = {};
 	private activeUserIds: string[] = [];
 
+	/** @hidden */
+	constructor(private name: string) {

```

Comment:
I think maybe the DiscreteAction class should get a reference back to the behavior that owns it.  That would get rid of the awkward plumbing in the `on` call below that doesn't feel right.  I had already felt like a reference back would have been good here, but I didn't initially need it before.  Seems like a good time to add it here.

Reply:
Are DiscreteActions specific to behaviors? They seem generic enough they could work for colliders and other kinds of things that generate events.

---
In reply to: [316826396](https://github.com/microsoft/mixed-reality-extension-sdk/pull/425#discussion_r316826396) [](ancestors = 316826396)

Reply:
Discrete actions are designed specifically for the behavior system.  Handling of trigger and collision events are a bit different and contain different data.  Adding in behavior reference here is the right thing to do I believe.

Reply:
If we decide to re-purpose this, we would likely have base class and subclass relationships for them anyways.

Reply:
How would grab work in this case? The grab DiscreteAction lives directly on Actor.

---
In reply to: [316832001](https://github.com/microsoft/mixed-reality-extension-sdk/pull/425#discussion_r316832001) [](ancestors = 316832001)

Reply:
That makes me wonder if grab should be moved to a GrabBehavior.

---
In reply to: [316840866](https://github.com/microsoft/mixed-reality-extension-sdk/pull/425#discussion_r316840866) [](ancestors = 316840866,316832001)

Reply:
Ah yes.  I forgot about grab.  Yeah.  I guess the `on` function will need to take those parameters.  Just seems awkward.  Grab was not a behavior intentionally as from the perspective of the SDK, we want grabbability to be a characteristic of actors all up.  Not whether they have a behavior on them or not.